### PR TITLE
Add Python crate and improve error handling

### LINKED ISSUE #20: Python bindings
Python bindings are needed for integration into pyQuil.

### LINKED ISSUE #21: Python bindings crash during libquilc's "init" call
TBD

### LINKED ISSUE #22: Improve error messages
Currently if any libquilc calls fail, we don't get an error message back. Rob &co suggested we implement a system where libquilc will store an error message and provide a new function `char* quil_last_error_message()` which queries the error message.

### PR DESCRIPTION
**Pull request by notmgsk**
This MR:

* adds a new `python` crate which provides bindings to build programs and chips, and compile programs (depends on an as-yet unmerged quilc change: https://github.com/quil-lang/quilc/pull/893)
* fixes the issue where symbols from `libquilc.so` were not visible to the `libsbcl.so` library
* improves errors (depends on an as-yet unmerged quilc change: https://github.com/quil-lang/quilc/pull/894)

Closes #20
Closes #21 
Closes #22

### CODE REVIEW

File: lib/build.rs [hunk 1]
```diff
@@ -15,6 +15,7 @@ fn main() {
     println!("cargo:rustc-link-search={}", quilc_library_path.display());
 
     println!("cargo:rustc-link-lib=quilc");
+    // println!("cargo:rustc-link-lib=sbcl");
```
Comment (notmgsk) [inline lib/build.rs:18]: remove

File: lib/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+            init(ptr);
+        }
+    })
+}
+
+/// A quilc chip specification
+#[derive(Clone, Debug)]
+pub struct Chip(chip_specification);
+
+unsafe impl Send for Chip {}
```
Comment (notmgsk) [inline lib/src/lib.rs:68]: These are required to play nicely with pyo3.

Comment (BatmanAoD) [reply lib/src/lib.rs:68]: What is `chip_specification, and how do we know it's safe to `Send`?

File: lib/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+            Ok(mut path) => {
+                path.push('\0');
+                path.as_bytes().to_vec()
+            }
+            Err(_) => b"libquilc.core\0".to_vec(),
+        };
+        let mut c_chars: Vec<i8> = bytes.iter().map(|c| *c as i8).collect();
+        let ptr = c_chars.as_mut_ptr();
+        unsafe {
+            init(ptr);
```
Comment (BatmanAoD) [inline lib/src/lib.rs:59]: I'm assuming `init` is some kind of FFI call into the Lisp runtime? This looks a little strange to me, with the duplicate env-var reading, manual iteration, etc.

To be more idiomatic and avoid the copying & manual iteration and `null`-appending, you should be able to just read the env var once, and convert the `OsString` into a `CString`:

```
let path = std::env::var("LIBQUILC_CORE_PATH").unwrap_or(OsString::from("libquilc.core"));
// do the path check
ptr = Cstring::new(path.into_vec()).into_raw();
```

Comment (notmgsk) [reply lib/src/lib.rs:59]: Still needs map & cast to `i8` unfortunately. I have used `CString` elsewhere in this MR but not here. Good call-out.

Comment (BatmanAoD) [reply lib/src/lib.rs:59]: `CString::into_raw` returns a pointer of `i8` type!

Comment (BatmanAoD) [reply lib/src/lib.rs:59]: (Some FFI trivia that's worth knowing: per the C standard, it's implementation defined whether `char` is signed or unsigned, so technically it shouldn't make any difference whether you pass `*mut i8` or `*mut u8` as your `char*` type.)

File: lib/src/lib.rs (hunk 2 above)
Comment (notmgsk) [reply lib/src/lib.rs:68]: A pointer to memory (which is not mutated) used by quilc.

Comment (BatmanAoD) [reply lib/src/lib.rs:61]: `unsafe` code should always have a comment explaining why it's valid, I think; could you please add a comment (here and on the other `unsafe impl`) explaining that the underlying memory is not expected to ever be mutated (at least while Python can hold a pointer to it)?

**Pull request merged**